src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs

```diff
@@ -162,38 +162,48 @@ private unsafe void WriteCore(ReadOnlySpan<byte> buffer)
+                // that accepts a CancellationToken is available.
+
+                // If a cancelable token is used and there's no data, issue a zero-length read so that
+                // we're asynchronously notified when data is available, and concurrently monitor the
+                // supplied cancellation token.  If cancellation is requested, we will end up "leaking"
+                // the zero-length read until data becomes available, at which point it'll be satisfied.
+                // But it's very rare to reuse a stream after an operation has been canceled, so even if
+                // we do incur such a situation, it's likely to be very short lived.
+                if (cancellationToken.CanBeCanceled)
                 {
                     cancellationToken.ThrowIfCancellationRequested();
-                    if (socket.Poll(timeout, SelectMode.SelectRead))
+                    if (socket.Available == 0)
                     {
-                        return ReadCore(destination.Span);
+                        Task<int> t = socket.ReceiveAsync(Array.Empty<byte>(), SocketFlags.None);
+                        if (!t.IsCompletedSuccessfully)
+                        {
+                            var cancelTcs = new TaskCompletionSource<bool>();
+                            using (cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).TrySetResult(true), cancelTcs))
+                            {
+                                if (t == await Task.WhenAny(t, cancelTcs.Task).ConfigureAwait(false))
+                                {
+                                    t.GetAwaiter().GetResult(); // propagate any failure
+                                }
+                                cancellationToken.ThrowIfCancellationRequested();
+
+                                // At this point there was data available.  In the rare case where multiple concurrent
+                                // ReadAsyncs are issued against the PipeStream, worst case is the reads that lose
+                                // the race condition for the data will end up in a non-cancelable state as part of
+                                // the actual async receive operation.
+                            }
+                        }
                     }
-                    timeout = Math.Min(timeout * 2, MaxTimeoutMicroseconds);
                 }
-            }
 
-            // The token wasn't cancelable, so we can simply use an async receive on the socket.
-            try
-            {
-                if (destination.TryGetArray(out ArraySegment<byte> buffer))
-                {
-                    return await socket.ReceiveAsync(buffer, SocketFlags.None).ConfigureAwait(false);
-                }
-                else
-                {
-                    // TODO #22608: Remove this terribly inefficient special-case once Socket.ReceiveAsync
-                    // accepts a Memory<T> in the near future.
```

Comment:
Q.: since #22608 is still open shouldn't this comment still be kept as a comment?

Reply:
I effectively just moved it up to here:
https://github.com/dotnet/corefx/pull/24302/files#diff-3aff29b58d0a45aaaba7664306a8a145R167